### PR TITLE
add claim/unclaim device, add user checking for getDevices() and rena…

### PR DESCRIPTION
**Pull request by AntonPuko**
…meDevice()

### CODE REVIEW

File: src/lib/controllers/DevicesController.js [hunk 1]
```diff
@@ -19,11 +19,40 @@ class DevicesController extends Controller {
...
+  @httpVerb('delete')
+  @route('/v1/devices/:deviceID')
+  async unclaimDevice(deviceID: string): Promise<*> {
+    try {
+      const userID = this.user.id;
+      await this._deviceRepository.unclaimDevice(deviceID, userID);
+
+      return this.ok({ ok: true });
+    } catch (exception) {
+      return this.bad(exception.message);
```
Comment (jlkalberer) [inline src/lib/controllers/DevicesController.js]: I'm thinking that we should move the `try/catch` to the `RouteConfig` if possible. It seems like all the controllers repeat the same logic over and over.

File: src/lib/controllers/DevicesController.js [hunk 2]
```diff
@@ -70,17 +102,17 @@ class DevicesController extends Controller {
         );
         return this.ok({ id: deviceID, status: 'Update started' });
       }
-console.log(postBody);
+
       // TODO not implemented yet
       // 3 flash device with precompiled binary
       if (postBody.file_type === 'binary' && this.request.files.file) {
-        this._deviceRepository.flashBinary(deviceID, this.request.files.file);
+        await this._deviceRepository.flashBinary(deviceID, this.request.files.file);
```
Comment (jlkalberer) [inline src/lib/controllers/DevicesController.js]: I'm not actually sure if we want to wait for this one. Flashing the device can take some time so I think that we might just want to immediately return.  You should test with the Particle cloud to make sure.

File: src/lib/controllers/DevicesController.js [hunk 3]
```diff
@@ -98,6 +130,7 @@ console.log(postBody);
         postBody,
       );
 
+      // TODO add userID checking
```
Comment (jlkalberer) [inline src/lib/controllers/DevicesController.js]: You should do this next.  That way the logic is in one place.. In DeviceRepository you do some logic there which I think could be moved to DeviceAttributeRepository.

File: src/lib/repository/DeviceRepository.js [hunk 4]
```diff
@@ -25,8 +25,49 @@ class DeviceRepository {
...
+      ownerID: userID,
+    };
+    return await this._deviceAttributeRepository.update(attributesToSave);
+  };
+
+  unclaimDevice = async (
+    deviceID: string,
+    userID: string,
+  ): Promise<DeviceAttributes> => {
+    const userDevicesAttributes = await this.getAll(userID);
```
Comment (jlkalberer) [inline src/lib/repository/DeviceRepository.js]: This should just be `getByID` with a `userID` filter.

File: src/lib/controllers/DevicesController.js (hunk 2 above)
Comment (AntonPuko) [inline src/lib/controllers/DevicesController.js]: I'm not sure yet, but as I know we should wait for `update started` status and then send ok response.

File: src/lib/HttpError.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// @flow
+
+class HttpError extends Error {
+  status: number;
+
+  constructor(message: string, status?: number = 400) {
+    super();
```
Comment (jlkalberer) [inline src/lib/HttpError.js]: `message` should be passed into super.

File: src/lib/RouteConfig.js [hunk 6]
```diff
@@ -94,18 +95,32 @@ export default (
...
+
+          try {
+            const result = mappedFunction.call(
+              controllerContext,
+              ...values,
+              body,
+            );
+
+            if (result.then) {
+              // eslint-disable-next-line no-shadow
```
Comment (jlkalberer) [inline src/lib/RouteConfig.js]: can this go away?

**Pull request merged**